Heartbeat desktop - implementation of heartbeat storage

AlmostMatt opened the pull request:
This is an initial implementation of storage for heartbeat desktop implementation

There are a number of TODOs for details that I am still working on, but the rough outline is there

AlmostMatt commented on app/src/heartbeat/heartbeat_storage_desktop.cc:
```diff
@@ -0,0 +1,125 @@
+  auto heartbeats_offset =
+      CreateLoggedHeartbeats(fbb, heartbeats_unique_ptr.get());
+  fbb.Finish(heartbeats_offset);
+  file.write((char*)fbb.GetBufferPointer(), fbb.GetSize());
```
The logic here is not good. All of the documentation and examples I could find for writing a flatbuffer to a disk had a pointer to the buffer of bytes. But I could not find a good way to go from the LoggedHeartbeats object to the buffer pointer, only the flat buffer builder has a convenient getter for the buffer pointer.

If there is a more convenient way to get the buffer pointer from the class I would be happy to use that.

chkuang-g replied:
@jonsimantov  Are you familiar about how this is done with Flatbuffer?

AlmostMatt replied:
I spoke a bit with Jon about flatbuffers and serializations. Seems that it is correct that the buffer pointer and size is needed to serialize, and LoggedHeartbeats is just a part of the buffer, not the whole thing.

Since LoggedHeartbeats is not too easy to modify or write, and a byte array is too implementation-specific, I am considering changing these methods to input and output a simple C struct that contains a last-logged field and a std::map from user agent to a list of date strings. This will mean a little bit of extra code to convert between the two, but the controller code would no longer have to deal with flatbuffer builders to modify the in-memory object.
In that case, the benefit of flatbuffers for serialization is mainly just that it allows us to verify the validity of the structure when it reads a byte buffer from a file.

chkuang-g replied:
It is good idea to free controller from handling flatbuffer.

Given that there is not change in this file, I assume your change will come in in the future?

AlmostMatt commented:
There are three future improvements that I would like to make but I think this change can be reviewed as-is.
- I would like to validate the date strings when reading from file
- I would like to lock the underlying files (using flock)
- I have a test for reading corrupted data, and it currently makes an assumption about the saved filename. It would be nice if there was a cleaner way to expose saved filename in tests.

chkuang-g reviewed (commented):
Also, did you have time to look into https://github.com/firebase/firebase-cpp-sdk/pull/897#discussion_r853552296?

AlmostMatt commented on app/src/heartbeat/heartbeat_storage_desktop.cc:
```diff
@@ -0,0 +1,125 @@
+  }
+
+  // TODO(almostmatt): does this need to be different between windows/mac/linux?
+  return app_dir + "/" + kHeartbeatFilename;
```
When I searched online it seemed like fstream might have internal logic to convert / and \ for windows. I haven't verified this yet, though in theory some of the added unit tests would confirm this if run on windows.

AlmostMatt replied:
Ok, I have confirmed that this does indeed work.
I wrote a simple script that uses fstream to write to "C:/Users/username/Documents/filename1" and also use fstream to write to "C:\\Users\\username\\Documents\\filename2", and ran the script on a windows device. Both files were created.

AlmostMatt replied:
It looks like the windows implementation of AppDataDir also uses forward slashes. I think that if we were to start using backslashes for windows implementation we would want to be consistent and update the existing AppDataDir function as well, at which point I think it would make sense to be a separate change from this heartbeats implementation.
I'm happy to do so if you would like.

chkuang-g commented on app/src/heartbeat/heartbeat_storage_desktop.cc:
```diff
@@ -0,0 +1,178 @@
+// Size is arbitrary, just making sure that there is a sane limit.
+static const int kMaxBufferSize = 1024 * 500;
+
+bool HeartbeatStorageDesktop::ReadTo(LoggedHeartbeats* heartbeats_output) {
```
http://go/cstyle#Inputs_and_Outputs

If this parameter is required, use `LoggedHeartbeats& heartbeats_output`

chkuang-g commented on app/src/heartbeat/heartbeat_storage_desktop.cc:
```diff
@@ -0,0 +1,178 @@
+  return *mutex_;
+}
+
+std::string GetFilename(std::string app_id, std::string* error) {
```
`const std::string& app_id`

chkuang-g commented on app/src/heartbeat/heartbeat_storage_desktop.cc:
```diff
@@ -0,0 +1,178 @@
+  return true;
+}
+
+bool HeartbeatStorageDesktop::Write(LoggedHeartbeats heartbeats) const {
```
`const LoggedHeartbeats& heartbeats` since this is for read only.

chkuang-g commented on app/src/heartbeat/heartbeat_storage_desktop.cc:
```diff
@@ -0,0 +1,178 @@
+}
+
+LoggedHeartbeats HeartbeatStorageDesktop::LoggedHeartbeatsFromFbs(
+    const LoggedHeartbeatsFbs* heartbeats_fbs) const {
```
If it is required, use `const LoggedHeartbeatsFbs&`

chkuang-g commented on app/src/heartbeat/heartbeat_storage_desktop.cc:
```diff
@@ -0,0 +1,178 @@
+}
+
+flatbuffers::FlatBufferBuilder HeartbeatStorageDesktop::LoggedHeartbeatsToFbs(
+    LoggedHeartbeats heartbeats_struct) const {
```
`const LoggedHeartbeats&` to prevent class copy

chkuang-g commented on app/src/heartbeat/heartbeat_storage_desktop.h:
```diff
@@ -0,0 +1,94 @@
+
+  // If the previous disk operation failed, contains additional details about
+  // the error; otherwise is empty.
+  const std::string& GetError() const { return error_; }
```
I am not so funded to this pattern.

1. If changing this field is guarded by a mutex, you should guard it when getting this property as well, since this can be changed from a different thread. Even this property is guarded, when the called try to access it, the error may not be relevant to what the called did.
2. It is unclear from an API level that `ReadTo()` and `Write()` would change this property and expect the called to retrieve error using a different API. 

I can think of a couple of approaches for this
1. Change API to something like `bool ReadTo(LoggedHeartbeats* heartbeats_output, std::string* error);`. This makes `error` arg optional and should be handled in the implementation.
2. Just log the error.  This can be done in a couple of ways. Each FirebaseApp instance has its own logger. https://github.com/firebase/firebase-cpp-sdk/blob/main/app/src/app_common.cc#L186. You can keep a reference to `App`, or the logger itself. Or you can create your own logger for an instance of Heartbeat. But I think former is better since it is easier to set verbosity level.

AlmostMatt replied:
Ok. For now I removed the mutex.
Deciding whether to rely on some new mutex, or scheduler, or just OS file locking is something that I think can be added while working on the controller.

AlmostMatt replied:
Ok following your later suggestion, I added Logger* as a parameter in the constructor for the class, and anywhere that it previously set error_, it will now call logger_->LogError.

AlmostMatt replied:
Another pattern for error output that I have seen is to have Read/Write return
absl::StatusOr<Type>

Then if it succeeds the return value is Type, and if it fails the return value is a Status which can have an error message (like absl::InvalidArgumentError("bad mode"))
But I don't see absl::Status being used much in Firebase C++ SDK

chkuang-g commented on app/src/heartbeat/heartbeat_storage_desktop.cc:
```diff
@@ -0,0 +1,178 @@
+const char kHeartbeatFilenamePrefix[] = "heartbeats-";
+
+// Returns the mutex that protects accesses to the storage file.
+Mutex& FileMutex() {
```
I think this mutex is better to be in class level.

That is, heartbeat storages for app1 and app2 are not necessarily writing to the same location. BUT it is also possible that they are write to the same location as well. Rare but possible.

Beside, static mutex is a sources of troubles. Since you did not write anything to destroy it, that is memory leak as well.

On the other hand, if you can guarantee that the access to this API are all serialized to the same thread, like using `schedular` class, the mutex is actually not required.

AlmostMatt replied:
Ok. For now I removed the mutex.
Deciding whether to rely on some new mutex, or scheduler, or just OS file locking is something that I think can be added while working on the controller.

chkuang-g commented:
Also, failed test
https://github.com/firebase/firebase-cpp-sdk/runs/6216165938?check_suite_focus=true

AlmostMatt commented on app/tests/heartbeat_storage_desktop_test.cc:
```diff
@@ -0,0 +1,210 @@
+}
+
+TEST_F(HeartbeatStorageDesktopTest, WriteAndReadWithSymbolsInAppId) {
+  std::string app_id = "start/\\?%*:|\"<>.,;=end";
```
Rather than exposing a static method that constructs a filename, I just provided a getter to the filename that is used by an instance of storage. I also updated the test to check the generated filename when the appid has symbols.

AlmostMatt commented on app/src/heartbeat/heartbeat_storage_desktop.h:
```diff
@@ -0,0 +1,70 @@
+
+  // Reads an instance of LoggedHeartbeats from disk into the provided struct.
+  // Returns `false` if the read operation fails.
+  bool ReadTo(LoggedHeartbeats& heartbeats_output);
```
When I make this output variable a non-const reference it gives a lint warning which is annoying, but I agree that the inputs/outputs documentation you linked does seem to suggest using non-const references for required output parameters.

chkuang-g replied:
agreed.  We should update the lint check here

AlmostMatt commented:
I made all of the recommended minor changes like pointer vs const reference.
I added Logger* as a parameter to the storage class. Errors are logged instead of an error_ string being updated.
I added a Getter to return the filename that is used by an instance of the storage class
I removed the Mutex and associated synchronization test. How synchronization will be handled will be decided when implementing the controller.

chkuang-g commented on app/src/heartbeat/heartbeat_storage_desktop.cc:
```diff
@@ -0,0 +1,174 @@
+
+}  // namespace
+HeartbeatStorageDesktop::HeartbeatStorageDesktop(const std::string& app_id,
+                                                 Logger* logger)
```
If `logger` is optional, you should do null-check whenever using it.

An easy solution is to make it required by using `const Logger& logger` and keep a reference to it.
Or you can implement functions like 
```
void HeartbeatStorageDesktop::LogWarning(const char* format, ...) {
if (logger_) {
  va_list list;
  va_start(list, format);
  logger_->LogWarning(format, list)
  va_end(list);
}
}
```
See  this https://github.com/firebase/firebase-cpp-sdk/blob/main/app/src/logger.cc#L23.

But I think this could be a bit too much trouble.
